test: sock: Enable all socket tests on Windows

davidgumberg opened the pull request:
Some `class Sock`  tests were previously disabled because Windows lacks [`socketpair(2)`](https://man7.org/linux/man-pages/man2/socketpair.2.html) which is used as a helper function in the socket tests, but is not strictly necessary or related to the `Socket` class under testing. This PR adds a `CreateSocketPair()` helper which creates a sender socket and receiver socket with a TCP connection, enabling these test cases for Windows. This also enables future tests that require more granular control over sockets than what `socketpair()` allows for, like using `setsockopt()` before connecting a socket.

This change is generally an improvement, but is also broken out of a [branch](github.com/davidgumberg/bitcoin/tree/2025-09-02-0xB10C-prefill-rebase) that does compact block prefilling up to the available bytes in the connection's current TCP window (see [delving post](https://delvingbitcoin.org/t/stats-on-compact-block-reconstructions/1052/34)). Creating connected socket pairs is useful for added tests in that branch that validate querying the current TCP window state, and without this change those tests don't run on Windows.

starkshade commented on src/test/sock_tests.cpp:
```diff
@@ -83,11 +83,46 @@ BOOST_AUTO_TEST_CASE(move_assignment)
 #ifndef WIN32 // Windows does not have socketpair(2).
 
-static void CreateSocketPair(int s[2])
+static std::pair<Sock, Sock> CreateSocketPair()
```
## ⚠️ Code Issue

**Lines:** 86-89

```cpp
static void CreateSocketPair(int s[2])
{
    BOOST_REQUIRE_EQUAL(socketpair(AF_UNIX, SOCK_STREAM, 0, s), 0);
}
```

The original `CreateSocketPair` function modified an array passed by reference. The new implementation returns a `std::pair<Sock, Sock>` by value, which is a better approach for managing socket resources, but the original function signature and implementation are removed, and the new implementation for non-Windows is now the primary one.

starkshade commented on src/test/sock_tests.cpp:
```diff
@@ -101,37 +136,23 @@ static void SendAndRecvMessage(const Sock& sender, const Sock& receiver)
-    delete sock1;
+        sock0moved = std::make_unique<Sock>(std::move(sock0));
+        sock1moved = std::make_unique<Sock>(INVALID_SOCKET);
+        *sock1moved = std::move(sock1);
```
## ⚠️ Code Issue

**Lines:** 147-151

```cpp
    int s[2];
    CreateSocketPair(s);

    Sock sock_send(s[0]);
    Sock sock_recv(s[1]);
```

Similar to the `wait` test case, this code initializes `Sock` objects directly from integer file descriptors. This bypasses RAII and could lead to resource management issues. Using `std::move` to transfer ownership is a more robust approach.

starkshade reviewed (commented):
The changes refactor the `CreateSocketPair` function to provide a cross-platform implementation and improve the handling of `Sock` objects using `std::unique_ptr` and move semantics, enhancing test robustness and memory management.

**Key Changes:**
- Introduced a Windows-compatible implementation for `CreateSocketPair` using TCP loopback connections, as `socketpair(2)` is not available on Windows.
- Modified `CreateSocketPair` to return `std::pair<Sock, Sock>` instead of modifying an integer array, promoting RAII and better ownership semantics.
- Replaced raw pointer usage (`new Sock(...)`) with `std::unique_ptr` in `send_and_receive` and `wait` test cases for safer memory management.
- Leveraged structured binding (`auto [sock0, sock1] = ...`) for cleaner initialization of socket pairs in test cases.
- Removed the `ifndef WIN32` block around the `CreateSocketPair` declaration, making it a single function with conditional logic inside.
- Ensured that `Sock` objects are properly moved and managed, especially after move assignment, in the `send_and_receive` test.

w0xlt reviewed (commented):
Concept ACK

davidgumberg committed:
test: sock: Enable socket pair tests on Windows

Adds a helper struct `socket_pair` for constructing and connecting TCP
sockets, this replaces socketpair() which is not available on Windows.
Making the socket pair TCP sockets instead of Unix sockets, and
separating socket creation from socket connection also enables more
detailed tests to be added in the future.

laanwj commented:
Concept ACK. Should we keep using `socketpair` on platforms that support it?

laanwj commented on src/test/sock_tests.cpp:
```diff
@@ -81,61 +81,75 @@ BOOST_AUTO_TEST_CASE(move_assignment)
+        // Get the address of the listener.
+        sockaddr_in bound{};
+        socklen_t blen = sizeof(bound);
+        BOOST_REQUIRE_EQUAL(receiver->GetSockName(reinterpret_cast<sockaddr*>(&bound), &blen), 0);
```
Might want to check `blen == sizeof(bound)` after this as well

laanwj commented:
> My gut reaction is that it would not be worth the extra code to do this, but I don't feel very strongly and am happy to add this if you think it's better.

Not 100% sure. As long as this is only used for the tests, like now, i don't mind it being like this. But if it were to be used for anything in the actual code, i think a socket pair should be a socket pair. Using a fallback work-around for functionality that's plainly supported seems strange.

laanwj reviewed (approved):
Code review ACK 15cefcd7d3729b3af59f63b1ba406b5dfb405e99

laanwj commented:
> Even though it's probably a bad idea to make changes in the present to suit a future use case that might never come about, I think it is generally useful to be able to create and connect the sockets separately in sock_tests.cpp and to be able to create TCP sockets like the application does.

If it's intentionally TCP then calling it `tcp_socket_pair` might be even better. So that no one decides to make this improvement in the meantime.

davidgumberg commented:
> calling it `tcp_socket_pair` might be even better

Thanks, fixed!

sedited reviewed (approved):
Isn't the usage of a socket pair kind of incidental to the test here? We do use unix sockets in a bunch of places now, but not through the socket class. From what I glean from the original PR, the purpose really is primarily to test the socket class itself (https://github.com/bitcoin/bitcoin/pull/20788/changes/615ba0eb96cf131364c1ceca9d3dedf006fa1e1c), so this change makes sense to me. I think it would be good to mention that in the PR description.

w0xlt reviewed (commented):
ACK 5dc3352be41a874272e92dd61f47bb08c386f523

w0xlt reviewed (commented):
reACK 0a8f5199048d56066bf87ea9e8f5b584b968140b

sedited reviewed (commented):
There is some simple style cleanup that could be done here: Using UpperCamelCase for clases, structs, and functions, dropping static methods in favour of a constructor, and dropping the seemingly unused `connect` bool:

<details>
<summary>Click to expand diff</summary>

```diff
diff --git a/src/test/sock_tests.cpp b/src/test/sock_tests.cpp
index 0cd5072758..2353b1b933 100644
--- a/src/test/sock_tests.cpp
+++ b/src/test/sock_tests.cpp
@@ -84 +84 @@ BOOST_AUTO_TEST_CASE(move_assignment)
-struct tcp_socket_pair {
+struct TcpSocketPair {
@@ -88 +88,3 @@ struct tcp_socket_pair {
-    static tcp_socket_pair create(bool connect)
+    TcpSocketPair()
+        : sender{std::make_unique<Sock>(CreateSocket())},
+        receiver{std::make_unique<Sock>(CreateSocket())}
@@ -90,6 +92 @@ struct tcp_socket_pair {
-        tcp_socket_pair socks{};
-        socks.sender = std::make_unique<Sock>(CreateSocket());
-        socks.receiver = std::make_unique<Sock>(CreateSocket());
-        if (connect) socks.connect_pair();
-
-        return socks;
+        Connect();
@@ -98 +95 @@ struct tcp_socket_pair {
-    void connect_pair()
+    void Connect()
@@ -122 +119 @@ struct tcp_socket_pair {
-    void send_and_receive()
+    void SendAndReceive()
@@ -136,2 +133,2 @@ BOOST_AUTO_TEST_CASE(send_and_receive)
-    tcp_socket_pair socks = tcp_socket_pair::create(/*connect=*/true);
-    socks.send_and_receive();
+    TcpSocketPair socks;
+    socks.SendAndReceive();
@@ -140,2 +137,2 @@ BOOST_AUTO_TEST_CASE(send_and_receive)
-    tcp_socket_pair socks_moved = std::move(socks);
-    socks_moved.send_and_receive();
+    TcpSocketPair socks_moved = std::move(socks);
+    socks_moved.SendAndReceive();
@@ -146 +143 @@ BOOST_AUTO_TEST_CASE(wait)
-    tcp_socket_pair socks = tcp_socket_pair::create(/*connect=*/true);
+    TcpSocketPair socks;
@@ -160 +157 @@ BOOST_AUTO_TEST_CASE(recv_until_terminator_limit)
-    tcp_socket_pair socks = tcp_socket_pair::create(/*connect=*/true);
+    TcpSocketPair socks;
```

</details>

sedited reviewed (approved):
ACK 9316d96240ff1bd4c98a823022655c44a04d1b20

davidgumberg commented on src/test/sock_tests.cpp:
```diff
@@ -82,61 +81,75 @@ BOOST_AUTO_TEST_CASE(move_assignment)
+    TcpSocketPair(const TcpSocketPair&) = delete;
+    TcpSocketPair& operator= (const TcpSocketPair&) = delete;
+    TcpSocketPair(TcpSocketPair&&) = default;
+    TcpSocketPair& operator= (TcpSocketPair&&) = default;
```
I think not, since `Sock` has it's copy constructor deleted the copy constructor for `TcpSocketPair` is also implicitly deleted:

https://github.com/bitcoin/bitcoin/blob/9316d96240ff1bd4c98a823022655c44a04d1b20/src/util/sock.h#L36-L39

I just use this as a primitive/self-documenting way to be 100% sure that I'm using the move constructor here:

https://github.com/bitcoin/bitcoin/blob/48b952cbb67e4a59e1f036375240bbb23ae21648/src/test/sock_tests.cpp#L115

without having to think about it very carefully, but if you prefer I can remove it, since it might be obvious to most readers that `Sock` is move-only.

w0xlt reviewed (commented):
reACK 9316d96240ff1bd4c98a823022655c44a04d1b20
